src/parse/session.rs

```diff
@@ -218,6 +218,15 @@ impl ParseSess {
         self.parse_sess.source_map().lookup_char_pos(pos).line
     }
 
+    // TODO(calebcartwright): Preemptive, currently unused addition
+    // that will be used to support formatting scenarios that take original
+    // positions into account
+    /// Determines whether a span begins on the same line where a previous span ends.
+    #[allow(dead_code)]
+    pub(crate) fn span_begins_same_line(&self, span: Span, prev: Span) -> bool {
```

Comment:
naming things is hard, 100% open to bikeshedding on this

Reply:
I actually think the name is fine!

Maybe consider naming the arguments `span` and `other`, or is `prev` chosen to suggest that we should always pass in a span that came from some previous location in the source text?

I could probably think of some scenarios related to comments where it might be nice to know if a comment span is trailing the content it's annotating. And I don't know if I'd necessarily call `span_begins_same_line` like:

```rust
span_begins_same_line(trailing_comment_span, span)
```

Maybe jumping off that comment span idea it might also be nice to have helper methods for `span_end_on_previous_line` and `span_start_on_next_line`, which might also be helpful in the context of determining pre and post comments for some `ast::Item` (`use` statements come to mind).

Reply:
hmm thinking about this more (and after looking at the code) the order is important since you're comparing the line position of `prev.hi()` with `span.lo()`. I'm wondering if `prev_span_ends_on_same_line` would be a more obvious name to communicate that there's an ordering to the span.

Depending on the context it might be more natural to write code that references future spans instead of previous spans so we might also consider adding a helper for `next_span_starts_on_same_line` that effectively does the same thing, but views the span's relationship from the other direction.

Reply:
Apologies I probably should've elaborated a bit more and not assumed all the dots were inherently connectable. 

High level, we're going to need some additional utility functions that facilitate peeking back into the source map for determining things like relative line positioning in the input source. 

Fundamentally, this is to support scenarios where we need/want to influence formatting based on the original, relative positions of elements in the input source. There are some near target scenarios, including some bugs we have with certain reorderings losing comment associations, as well as some new formatting options I want to make available for chains and call args. 

Those are the use cases which will be subsequently implemented with their own logic that will leverage these more trivial/binary utility functions as needed. However, I don't want any of the general purpose utility functions to be overly coupled to any specific scenario.

While I do anticipate that we will likely have several such utility functions when all is said and done, at the moment/in this PR I'm just focused on the "does span A start on the same line as the line where span B ends" one. You raise a good point about the parameter name so I'll think on that for a bit as it's not strictly a next/previous thing. Remember that with a tree structure there's all kinds of span combination cases where such a function will, and should return true, but which wouldn't be particularly useful at the call sites.

For example, consider a simple local statement that has an initializer expression, all of which can fit on one line:

```rust
fn main() {
    let x = 2;
}
```

We could compare the spans of the pattern, the initializer, or the entire statement (which encloses all of them), by using them interchangably in any arg position and the function would (correctly) return true in all cases. So it's really not a matter of the spans being sequential/abutting nor do they need to be "unrelated" in the sense that there's not a parent/child relationship between the containing nodes; it's really just a matter of callers utilizing the check with whatever pair(s) of spans make sense in their context.

I also want to avoid having the function name become a run-on sentence on its own, so hoping for something succinct that, along with docs, can clearly convey the "does span A start on the same line as the line where span B ends" message

Reply:
Just going to throw a bunch of names out there. Let me know if any resonate with / inspire you:
* `span_begins_same_line` (for context here's the original name)
* `begins_on_final_line`
* `starts_on_end_line`
* `spans_share_same_line`
* `spans_occupy_same_line`
* `adjacent_spans`
* `start_on_others_end`
* `occupies_others_end_line`
* `found_on_same_line`
* `overlapping_hi_lo_lines`
* `overlapping_lines`
* `intersecting_hi_lo_lines`
* `same_hi_lo_lines`
* `start_intersects_end`